cmd/appcatalogs.go

```diff
@@ -0,0 +1,194 @@
+	giantSwarmCatalogURL           = "https://giantswarm.github.io/giantswarm-catalog/index.yaml"
+	giantSwarmPlaygroundCatalogURL = "https://giantswarm.github.io/giantswarm-playground-catalog/index.yaml"
+
+	// Annotation keys
+	teamAnnotation = "application.giantswarm.io/team"
+)
+
+var defaultCatalogURLs = []string{
+	clusterCatalogURL,
+	controlPlaneCatalogURL,
+	defaultCatalogURL,
+	giantSwarmAzureCatalogURL,
+	giantSwarmCatalogURL,
+	giantSwarmPlaygroundCatalogURL,
+}
+
+func init() {
+	appCatalogsCmd.PersistentFlags().StringSliceP("url", "", defaultCatalogURLs, "App catalog urls")
+}
+
+func runAppCatalogs(cmd *cobra.Command, args []string) {
+	urls, err := cmd.PersistentFlags().GetStringSlice("url")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	path, err := cmd.Root().PersistentFlags().GetString("output")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	componentExporter := export.New(export.Config{TargetPath: path + "/components.yaml"})
+
+	entriesCount := 0
+	apps := make(map[string]int)
+
+	// Iterate over catalogs
+	for _, url := range urls {
+		fmt.Printf("Reading catalog %s\n", url)
+
+		index, err := appcatalog.LoadFromURL(url)
+		if err != nil {
+			log.Fatalf("Error loading app catalog from %s: %s", url, err)
+		}
+
+		log.Printf("Catalog generated at %s", index.Generated)
+
+		// Iterate over apps in catalog
+		for appName := range index.Entries {
+			entriesCount++
+			if _, ok := apps[appName]; ok {
+				// App already seen, skip
+				continue
+			}
+
+			if len(index.Entries[appName]) < 1 {
+				log.Printf("App %s has no releases", appName)
+				continue
+			}
+
+			apps[appName]++
+
+			component, err := componentFromCatalogEntry(index.Entries[appName][0])
+			if err != nil {
+				log.Printf("ERROR: Could not create component entity. %v", err)
+				continue
+			}
+
+			e := component.ToEntity()
+			err = componentExporter.AddEntity(e)
+			if err != nil {
+				log.Fatalf("Error: %v", err)
+			}
+
+		}
+	}
+
+	log.Printf("Collected %d unique apps in %d entries", len(apps), entriesCount)
+
+	err = componentExporter.WriteFile()
+	if err != nil {
+		log.Fatalf("Error writing components: %v", err)
+	}
+	log.Printf("Wrote file %s", componentExporter.TargetPath)
+}
+
+// Populates a catalog.Component from an appcatalog.Entry
+func componentFromCatalogEntry(entry appcatalog.Entry) (*catalog.Component, error) {
+	// owner team
+	team := "group:giantswarm/unspecified"
+	if teamName, ok := entry.Annotations[teamAnnotation]; ok {
+		// Adding Giant Swarm product team prefix
+		if !strings.HasPrefix(teamName, "team-") {
+			teamName = "team-" + teamName
+		}
+
+		team = "group:giantswarm/" + teamName
+	}
+
+	// release time
+	releaseTime, err := time.Parse(time.RFC3339, entry.Created)
+	if err != nil {
+		return nil, err
+	}
+
+	githubSlug := detectGitHubSlug(&entry)
+	if githubSlug == "" {
+		return nil, fmt.Errorf("could not detect GitHub slug for app %s in app metadata", entry.Name)
+	}
+
+	component, err := catalog.NewComponent(entry.Name,
```

Comment:
Nice, interesting structure. Never did quite something like that on go before. Good to know you can do it like this.

Reply:
It's called "options pattern". https://dev.to/kittipat1413/understanding-the-options-pattern-in-go-390c